src/command.rs

```diff
@@ -0,0 +1,306 @@
+        .await
+        .wrap_err("Failed to create release")?;
+
+    if dry_run {
+        tracing::warn!(":: aborting upload due to dry run");
+        return Ok(());
+    }
+
+    artifactory.publish(package, repository).await?;
+
+    Ok(())
+}
+
+/// Installs dependencies
+pub async fn install(credentials: Credentials) -> eyre::Result<()> {
+    let artifactory = {
+        let Some(artifactory) = credentials.artifactory else {
+            eyre::bail!(
+                "Unable to install artifactory dependencies, please login using `buffrs login`"
+            );
+        };
+
+        Artifactory::from(artifactory)
+    };
+
+    let manifest = Manifest::read().await?;
+
+    let lockfile = Lockfile::read_or_default().await?;
+
+    let dependency_tree =
+        DependencyGraph::from_manifest(&manifest, &lockfile, &artifactory).await?;
+
+    let mut locked = Vec::new();
+
+    #[async_recursion]
+    async fn process_recursively(
+        name: &PackageName,
+        tree: &DependencyGraph,
+        locked: &mut Vec<LockedPackage>,
+        prefix: String,
+    ) -> eyre::Result<()> {
+        let resolved = tree
+            .get(name)
+            .wrap_err("Dependency missing from dependency tree")?;
+
+        PackageStore::unpack(&resolved.package).await?;
+
+        tracing::info!(
+            "{} installed {}@{}",
+            if prefix.is_empty() { "::" } else { &prefix },
+            name,
+            resolved.package.version()
+        );
+
+        locked.push(resolved.package.lock(resolved.repository.clone()));
+
+        for (index, dependency) in resolved.depends_on.iter().enumerate() {
+            let tree_char = if index + 1 == resolved.depends_on.len() {
+                '┗'
+            } else {
+                '┣'
+            };
+
+            let new_prefix = format!(
+                "{} {tree_char}",
+                if prefix.is_empty() { "  " } else { &prefix }
+            );
+            process_recursively(dependency, tree, locked, new_prefix).await?;
+        }
+
+        Ok(())
+    }
+
+    for dependency in manifest.dependencies {
+        process_recursively(
+            &dependency.package,
+            &dependency_tree,
+            &mut locked,
+            String::new(),
+        )
+        .await?;
+    }
+
+    let lockfile = Lockfile::from_iter(locked.into_iter());
+    lockfile.write().await?;
+
+    Ok(())
+}
+
+/// Uninstalls dependencies
+pub async fn uninstall() -> eyre::Result<()> {
+    PackageStore::clear().await
+}
+
+/// Generate bindings for a given language
+pub async fn generate(language: Language) -> eyre::Result<()> {
+    generator::generate(language)
+        .await
+        .wrap_err_with(|| format!("Failed to generate language bindings for {language}"))?;
+
+    Ok(())
+}
+
+/// Logs you in for a registry
+pub async fn login(mut credentials: Credentials, url: url::Url) -> eyre::Result<()> {
+    let mut cfg = ArtifactoryConfig::new(url)?;
+
+    let password = {
```

Comment:
you might want to use a library which avoids putting the token in stdout as it's typed in?

Reply:
#63